### ISSUE TITLE
Theoretical issue in unknown_toUtf8()

### ISSUE THREAD

**Issue by bnason-nf** [libexpat/libexpat#9]
(migrating this from SourceForge)

Hi,

I ran the clang static analyzer on expat 2.2.0 and it reported this potential issue:

Logic error: Assigned value is garbage or undefined (expat/expat/lib/xmltok.c:1414)
expat/expat/lib/xmltok.c

```
  1400          int c = uenc->convert(uenc->userData, *fromP);
  1401          n = XmlUtf8Encode(c, buf);
  1402          if (n > toLim - *toP)
  1403            return XML_CONVERT_OUTPUT_EXHAUSTED;
  1404          utf8 = buf;
  1405          *fromP += (AS_NORMAL_ENCODING(enc)->type[(unsigned char)**fromP]
  1406                     - (BT_LEAD2 - 2));
  1407        }
  1408        else {
  1409          if (n > toLim - *toP)
  1410            return XML_CONVERT_OUTPUT_EXHAUSTED;
  1411          (*fromP)++;
  1412        }
  1413        do {
  1414          *(*toP)++ = *utf8++;
                          ^ Logic error: Assigned value is garbage or undefined
  1415        } while (--n != 0);
  1416      }
  1417    }
  1418    
  1419    static enum XML_Convert_Result PTRCALL
  1420    unknown_toUtf16(const ENCODING *enc,
  1421                    const char **fromP, const char *fromLim,
  1422                    unsigned short **toP, const unsigned short *toLim)
  1423    {
  1424      const struct unknown_encoding *uenc = AS_UNKNOWN_ENCODING(enc);
  1425      while (*fromP < fromLim && *toP < toLim) {
  1426        unsigned short c = uenc->utf16[(unsigned char)**fromP];
  1427        if (c == 0) {
  1428          c = (unsigned short)
```
Entering loop body
expat/expat/lib/xmltok.c

```
  1389    {
  1390      const struct unknown_encoding *uenc = AS_UNKNOWN_ENCODING(enc);
  1391      char buf[XML_UTF8_ENCODE_MAX];
  1392      for (;;) {
            ^ Entering loop body
  1393        const char *utf8;
  1394        int n;
  1395        if (*fromP == fromLim)
```

Assuming 'n' is equal to 0
expat/expat/lib/xmltok.c

```
  1396          return XML_CONVERT_COMPLETED;
  1397        utf8 = uenc->utf8[(unsigned char)**fromP];
  1398        n = *utf8++;
  1399        if (n == 0) {
                  ^ Assuming 'n' is equal to 0
  1400          int c = uenc->convert(uenc->userData, *fromP);
  1401          n = XmlUtf8Encode(c, buf);
  1402          if (n > toLim - *toP)
```

Assigned value is garbage or undefined
expat/expat/lib/xmltok.c

```
  1411          (*fromP)++;
  1412        }
  1413        do {
  1414          *(*toP)++ = *utf8++;
                          ^ Assigned value is garbage or undefined
  1415        } while (--n != 0);
  1416      }
  1417    }
```

The reason for the "garbage or undefined" is not completely clear to me, but I think it's saying that if n ends up being zero, then the do/while loop would behave very badly. Adding an "if (n > 0)" check around the do/while makes clang happy, but I'm not sure that's the best way to address it.

Thanks,
Benbuck

**Comment by RMJ10** [comment]
There are actually two potential problems here.

The initial value of "n" as assigned on line 1398 is something `XmlInitUnknownEncoding()` generates, so will be well behaved.  "n == 0" here means that the byte being processed is the start of a multi-byte sequence decoded by calling the user-supplied `uenc->convert()` function.  Our first problem is that there is no check that `uenc->convert()` actually exists.  It would be simple and cheap to add a check in `XmlInitUnknownEncoding()` to ensure that the convert function pointer is not NULL if there are any table entries that will need converting.

Assuming there is a convert function, can it provoke `XmlUtf8Encode()` into returning 0?  Yes, it's even documented.  If the convert function returns -1 (or indeed any negative number), this is taken to mean we have a malformed character.  When this happens through the mapping table, we return a NUL (0x00) as the conversion.  We should probably do the same here; it would be simplest to modify `XmlUtf8Encode()` to stick a 0 in the buffer and return 1 rather than returning 0 as it currently does.

**Comment by RMJ10** [comment]
Things are a little more complicated than I suggested, of course.

Commit 49308b7af63d4105339f40e0fb744ebfc217d0e7 on RMJ10/libexpat coverage branch demonstrates what happens when you need to convert but don't have a conversion function; the library segfaults.  Commit 6fdfe7a651d99d91f1752fed55908cf2cca9ce9a has a candidate fix for that.

However I was quite wrong about the convert function being able provoke `XmlUtf8Encode()` into returning 0.  What I wrote above was correct, but I failed to notice that long before that, the input string is tested for validity.  If the conversion function returns -1, it will cause an `XML_ERROR_INVALID_TOKEN` to be returned by an earlier stage of the parse, so `unknown_toUtf8()` never actually gets called.  Commit 0eeeeb9ec85474f18227a6c74820614f06c05f7f has a test demonstrating this.

So this suspicious code is in fact safe (modulo lacking a converter function).  If we really want to make clang's static analysis tool happy, we could add `if (n<0) return XML_CONVERT_COMPLETED` or similar after the call to `XmlUtf8Convert()`.  @hartwork do you have an opinion on this?

**Comment by hartwork** [comment]
> Our first problem is that there is no check that `uenc->convert()` actually exists.  It would be simple and cheap to add a check in `XmlInitUnknownEncoding()` to ensure that the convert function pointer is not NULL if there are any table entries that will need converting.
> [..]
> Commit 49308b7af63d4105339f40e0fb744ebfc217d0e7 on RMJ10/libexpat coverage branch demonstrates what happens when you need to convert but don't have a conversion function; the library segfaults.  Commit 6fdfe7a651d99d91f1752fed55908cf2cca9ce9a has a candidate fix for that.

That fix is on `master` as fc06a529a3283c9b0ec5e253e98f4c23d3b9caa8 by now.
I need more time to think about the remainder of this ticket.

**Comment by hartwork** [comment]
Turned out to be a false positive, silenced by 448d79c02e1b670ca69a7606c048165ae8126076 now. Thanks for the report!
